### PR TITLE
Add watch~unwatch feature to project

### PR DESCRIPTION
**Pull request by AnkushMalik**
**In this PR :** 
- Add addition relationship between user and project model , i.e. project_follow.rb .
- Add watch-unwatch buttons in view & follow-unfollow methods in project controller .
- Add modal for showing project's watchers .
---
screen shot 2018-06-20 at 9 59 38 pm

---
screen shot 2018-06-20 at 10 00 10 pm

### CODE REVIEW

File: spec/factories/project_follows.rb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+FactoryBot.define do
```
Comment (DavidKang) [inline spec/factories/project_follows.rb]: Please remove this file from this commit

File: spec/models/project_follow_spec.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+require 'rails_helper'
```
Comment (DavidKang) [inline spec/models/project_follow_spec.rb]: Please remove this file from this commit

File: app/controllers/projects_controller.rb [hunk 3]
```diff
@@ -196,6 +196,16 @@ def delete_episode
     redirect_to project_path(nil, @project), notice: "Removed hackweek #{@episode.name}"
   end
 
+  def follow
```
Comment (DavidKang) [inline app/controllers/projects_controller.rb]: The actions `follow` and `unfollow`, sounds me that they have to be in a specific controller with CRUD actions.

File: app/views/projects/_project_watch.html.haml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    = link_to unfollow_project_path(episode, project), class: 'btn btn-default watch_action' do
+      %i.fa.fa-eye-slash
+      Unwatch
+  - else
+    = link_to follow_project_path(episode ,project), class: 'btn btn-default watch_action' do
+      %i.fa.fa-eye
+      Watch
+  = link_to "#{project.project_followers.count}", '#', class: 'btn btn-default', data: { toggle: 'modal', target: '#watchersModal' }
+
+.modal.fade{id: "watchersModal","aria-labelledby" => "watchersModalLabel", role: "dialog", tabindex: "-1"}
```
Comment (DavidKang) [inline app/views/projects/_project_watch.html.haml]: Use single quotes for strings.

Comment (DavidKang) [inline app/views/projects/_project_watch.html.haml]: `.modal.fade{id: "watchersModal", ...}` could be written as: `#watchers-modal.modal.fade{ .... }`

Comment (DavidKang) [inline app/views/projects/_project_watch.html.haml]: Ids and classes should be written as: `watchers-modal` and not `watchersModal` as in javacript.

File: app/views/projects/_project_watch.html.haml [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+  .modal-dialog{role: "document"}
```
Comment (DavidKang) [inline app/views/projects/_project_watch.html.haml]: use single quotes please

File: app/views/projects/_project_watch.html.haml [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+      .modal-header
+        %button.close{'aria-label' => 'Close', 'data-dismiss' => 'modal', type: 'button'}
+          %span{'aria-hidden' => 'true'} ×
+        %h2#watchersModalLabel.modal-title Watchers of #{project.title}
+      .modal-body
+        - project.project_followers.each do |follower|
+          = link_to user_path(follower) do
+            .project_watcher_box
+              .col-md-4
+                = image_tag(follower.gravatar_url(:size => "40"), class: "media-object img-rounded", alt: follower.name, title: follower.name)
```
Comment (DavidKang) [inline app/views/projects/_project_watch.html.haml]: same here

**Comment by hennevogel** [comment]
Hm, what is the actual feature?

**Comment by AnkushMalik** [comment]
@hennevogel , the feature is simply to follow a project(just like in github). 
Currently there is no way to follow a project/idea in Hackweek by which project members (originator + project.users) can be notified about the activities going on in the projects they are interested in. Even if the user is not in the project member he will be notified through notifications and emails about the project activities, if he ‘watches’(ie is interested in) the respective project.

**Comment by hennevogel** [comment]
>  Even if the user is not in the project member he will be notified through notifications and emails about the project activities

And where is that functionality?

**Comment by AnkushMalik** [comment]
In this PR, i only implemented the follow/unfollow of projects.
I will implement notifications/email feature in my next two PRs

File: config/routes.rb [hunk 7]
```diff
@@ -40,6 +40,12 @@
         post 'keyword', to: "projects#add_keyword"
         delete 'keyword', to: "projects#delete_keyword", constraints: { keyword: /[^\/]+/ }
       end
+      resources :project_follows, only: [] do
+        member do
+          get '/create', to: 'projects/project_follows#create'
```
Comment (DavidKang) [inline config/routes.rb]: Should be `post` instead of `get`

File: config/routes.rb [hunk 8]
```diff
@@ -40,6 +40,12 @@
...
+          get '/destroy', to: 'projects/project_follows#destroy'
```
Comment (DavidKang) [inline config/routes.rb]: Should be `delete` instead of `get`

File: app/controllers/projects/project_follows_controller.rb [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    current_user.project_followings << @project
+    redirect_to project_path(@episode, @project), notice: "You are now watching #{@project.title}"
+  end
+
+  def destroy
+    current_user.project_followings.delete @project
+    redirect_to project_path(@episode, @project), notice: "You have stopped watching #{@project.title}"
+  end
+
+  private
```
Comment (DavidKang) [inline app/controllers/projects/project_follows_controller.rb]: Please, add a new line

File: spec/controllers/projects/project_follows_controller_spec.rb [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+  let(:project) { create :project }
+
+  before :each do
+    sign_in user
+  end
+
+  describe 'GET create' do
+    it 'creates a unique project_follow' do
+      expect {
+        get :create, project_id: project.id, episode: episode, id: project.id
```
Comment (DavidKang) [inline spec/controllers/projects/project_follows_controller_spec.rb]: `id: project.id`? why? :thinking:

File: spec/controllers/projects/project_follows_controller_spec.rb [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+      }.to change(ProjectFollow, :count).by(1)
+      expect {
+        get :create, project_id: project.id, episode: episode, id: project.id
```
Comment (DavidKang) [inline spec/controllers/projects/project_follows_controller_spec.rb]: same here

File: spec/controllers/projects/project_follows_controller_spec.rb [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+      }.to change(ProjectFollow, :count).by(0)
+    end
+  end
+  describe 'GET destroy' do
+    it 'deletes a project_follow entry' do
+      get :create, project_id: project.id, episode: episode, id: project.id
```
Comment (DavidKang) [inline spec/controllers/projects/project_follows_controller_spec.rb]: Better use a factory for that. :smile_cat:

File: spec/controllers/projects/project_follows_controller_spec.rb [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+      expect {
+        get :destroy, project_id: project.id, episode: episode, id: project.id
```
Comment (DavidKang) [inline spec/controllers/projects/project_follows_controller_spec.rb]: `id: project.id`? why? :thinking:

File: spec/controllers/projects/project_follows_controller_spec.rb (hunk 12 above)
Comment (AnkushMalik) [reply spec/controllers/projects/project_follows_controller_spec.rb]: need some help with this :|
factory for project_follow is created but the delete request in this method doesnt destroy that project_follow entry.

File: app/controllers/projects/project_follows_controller.rb (hunk 9 above)
Comment (AnkushMalik) [reply app/controllers/projects/project_follows_controller.rb]: added

**Pull request merged**